### PR TITLE
WebMercatorViewport support external projectionMatrix

### PR DESCRIPTION
**Pull request by felixpalmer**
Implementation of https://github.com/visgl/deck.gl/issues/5929

### CODE REVIEW

**Comment by felixpalmer** [comment]
Proposal for changing WebMercatorViewport construction. If you are happy with this approach I will go ahead and update the docs & add tests

**Comment by Pessimistress** [comment]
Can you not achieve this by offsetting `zoom`?

**Comment by Pessimistress** [comment]
I understand that you want to change camera position, however modifying scale does not just change camera z, but also x and y scaling.

https://github.com/uber-web/math.gl/blob/master/modules/web-mercator/src/web-mercator-utils.js#L176

https://github.com/uber-web/math.gl/blob/master/modules/web-mercator/src/web-mercator-utils.js#L183

`scale = 2^zoom` is used extensively across this code base in shaders and e.g. the TileLayer. Modifying it only in some places will result in many unintentional breakages.

If you just need to control camera position and fov separately, then add a fovy parameter that overrides the fovy calculation from altitude?

**Comment by felixpalmer** [comment]
@Pessimistress how is this?
- I named the parameter `fov` and not `fovy` as `fovy` conflicts with a prop in `modules/core/src/views/view.js `
- `this.altitude` needs to use the same "altitude" as the projection matrix otherwise deck.gl completely crashes (tab hangs)
- As background, I wrote up the approach I took in calculating the matrices to match Google Maps [here](https://github.com/visgl/deck.gl/blob/felix/google-vector/docs/developer-guide/base-map-viewport-alignment.md)

**Comment by Pessimistress** [comment]
> fovy conflicts with a prop in modules/core/src/views/view.js

It should be `fovy`. It is exactly what that prop intends to be. `WebMercatorViewport` just ignored it before.

**Comment by felixpalmer** [comment]
There's an issue with that:

https://github.com/visgl/deck.gl/blob/203a658c840586fa296a91210e24737ef46fae7f/modules/core/src/views/view.js#L154

`viewState` is overwritten with `this.props` so the default `fov` of 50 is picked up from the `View`. Is it OK to switch the precedence of the props, like so:

```javascript
    return new ViewportType({...this.props, ...viewState, ...viewportDimensions}); 
```

`fovy` seems to be the only conflicting prop

**Comment by Pessimistress** [comment]
No, the override order is an important feature. `fovy` is only set to 50 as a fall back in the constructor. You can change the default in the `MapView` constructor to `null` before passing it to the parent class.

**Comment by felixpalmer** [comment]
OK, thanks for the pointer. The problem is that the default `fov` of the View class will be set to 50 even if I don't change it, so I can't use its presence to update the projectionMatrix. Instead, I've come up with another way, which is to simply use the `projectionMatrix`, which is `null` by default

**Comment by Pessimistress** [comment]
Do you still need to make a change then? If you are supplying your own `projectionMatrix` then everything already work as is?

**Comment by felixpalmer** [comment]
If I don't update `this.altitude` I'm seeing the issue I mentioned above with deck.gl freezing when the pitch is increased to 63 degrees

**Comment by Pessimistress** [comment]
What causes the crash? Infinite recursive updates? Any console error? It needs to be explained otherwise future maintainers may very well break this again.

**Comment by felixpalmer** [comment]
`getBounds` passes `this.altitude` to math.gl

https://github.com/visgl/deck.gl/blob/0f65a4a80b21df6e53e1b9874c01771d9fdfd8a3/modules/core/src/viewports/web-mercator-viewport.js#L182

https://github.com/uber-web/math.gl/blob/d699d26ab39f1d415b92aa69247d735efa8e56bc/modules/web-mercator/src/get-bounds.js#L20

This then uses the altitude to obtain the field of view and perform an unprojection. Thus we need the altitude to match that of the projectionMatrix.

The crash is then due an invalid bounds object in `get subViewports()`

https://github.com/visgl/deck.gl/blob/0f65a4a80b21df6e53e1b9874c01771d9fdfd8a3/modules/core/src/viewports/web-mercator-viewport.js#L138-L147

As the projection is wrong the `minOffset` generated is -1706482542 and `maxOffset` is 1706482541 which leads the method to start creating a huge number of subviewports, each of which also creates more subviewports.

**Comment by Pessimistress** [comment]
`this.altitude` should always reflect `opts.altitude`. Swapping the value out after it's used for initialization is a hack and prone to future issues. Here is my proposal:

- add a field `this.fovy` to `WebMercatorViewport`, calculated from either `altitude` (default case) or `projectionMatrix` (this use case)
- math.gl's `getBounds` should look for `fovy` if defined and fall back to using `altitude` in its calculations

**Comment by felixpalmer** [comment]
I like this approach. I will create a PR in math.gl and then update here

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 1]
```diff
@@ -68,12 +71,13 @@ export default class WebMercatorViewport extends Viewport {
...
     // TODO - just throw an Error instead?
     altitude = Math.max(0.75, altitude);
+    const altitudeFromProjection = projectionMatrix && projectionMatrix[5] / 2;
 
     const {fov, aspect, focalDistance, near, far} = getProjectionParameters({
       width,
       height,
       pitch,
-      altitude,
+      altitude: projectionMatrix ? altitudeFromProjection : altitude,
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:80]:
```suggestion
      altitude: projectionMatrix ? projectionMatrix[5] / 2 : altitude,
```

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 2]
```diff
@@ -73,7 +76,7 @@ export default class WebMercatorViewport extends Viewport {
       width,
       height,
       pitch,
-      altitude,
+      altitude: projectionMatrix ? projectionMatrix[5] / 2 : altitude,
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:79]: I thought this was deprecated? Let's spell out the underlying logic here:

```js
const fov = altitudeToFov(projectionMatrix ? projectionMatrix[5] / 2 : altitude);

getProjectionParameters({fov, ...});
```

**Review by Pessimistress — APPROVED** [review]

File: modules/core/src/viewports/web-mercator-viewport.js (hunk 2 above)
Comment (felixpalmer) [reply modules/core/src/viewports/web-mercator-viewport.js:79]: Fixed, but requires a new math.gl release as these function were not exported

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 3]
```diff
@@ -109,7 +112,7 @@ export default class WebMercatorViewport extends Viewport {
 
       // projection matrix parameters
       orthographic,
-      fovyRadians: fov,
+      fovyRadians,
       aspect,
       focalDistance,
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:114]: I think `focalDistance` always needs to be set to `altitude`. You can test its effect with sizeUnits/radiusUnits/widthUnits: 'pixels'.

Comment (felixpalmer) [reply modules/core/src/viewports/web-mercator-viewport.js:114]: Even if I change that the size of the points is incorrect (they are too large by a factor of about 2.25) when I test it against the Google projection. `2.2553542518310286` I recognize as `fovyToAltitude(fovy)` for that projection (while `altitude` is `1`, so my feeling is that there is somewhere in deck.gl that making the assumption that `fovyToAltitude(fovy) === altitude` but I haven't yet found where

Comment (felixpalmer) [reply modules/core/src/viewports/web-mercator-viewport.js:114]: I've set focalDistance to always equal altitude, as (as per offline discussion) it turns out we can make this work with the Google matrices if we adjust the near plane

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 4]
```diff
@@ -69,14 +71,16 @@ export default class WebMercatorViewport extends Viewport {
     // TODO - just throw an Error instead?
     altitude = Math.max(0.75, altitude);
 
-    const {fov, aspect, focalDistance, near, far} = getProjectionParameters({
+    const fovy = altitudeToFovy(projectionMatrix ? projectionMatrix[5] / 2 : altitude);
+    const {aspect, fov: fovyRadians, near, far} = getProjectionParameters({
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:75]: Only run this if `projectionMatrix` is not supplied?

```js
let fovy;
if (projectionMatrix) {
  fovy = altitudeToFovy(projectionMatrix[5] / 2);
} else {
  fovy = altitudeToFovy(altitude);
  projectionMatrix = getProjectionMatrix(...);
}
```

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 5]
```diff
@@ -109,7 +113,7 @@ export default class WebMercatorViewport extends Viewport {
 
       // projection matrix parameters
       orthographic,
-      fovyRadians: fov,
+      fovyRadians,
       aspect,
       focalDistance,
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:114]: Nit: `focalDistance: altitude`

File: modules/core/src/viewports/web-mercator-viewport.js (hunk 4 above)
Comment (felixpalmer) [reply modules/core/src/viewports/web-mercator-viewport.js]: I've done this refactor, but with a slightly different approach so as to support the orthographic code path

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 6]
```diff
@@ -108,12 +126,7 @@ export default class WebMercatorViewport extends Viewport {
       zoom,
 
       // projection matrix parameters
-      orthographic,
-      fovyRadians: fov,
-      aspect,
-      focalDistance,
-      near,
-      far
+      ...projectionParameters
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:129]: focalDistance is not a projection parameter. Move it here?

File: modules/core/src/viewports/web-mercator-viewport.js [hunk 7]
```diff
@@ -69,14 +71,30 @@ export default class WebMercatorViewport extends Viewport {
...
-      farZMultiplier
-    });
+    let fovy;
+    let projectionParameters;
+    if (projectionMatrix) {
+      fovy = altitudeToFovy(projectionMatrix[5] / 2);
+      projectionParameters = {focalDistance: altitude, projectionMatrix};
+    } else {
+      fovy = altitudeToFovy(altitude);
+      const {aspect, fov: fovyRadians, near, far} = getProjectionParameters({
```
Comment (Pessimistress) [inline modules/core/src/viewports/web-mercator-viewport.js:81]: Perf: you can avoid destructing/constructing objects here:

```js
let projectionParameters = null;
if (projectionMatrix) {
  fovy = altitudeToFovy(projectionMatrix[5] / 2);
} else {
  fovy = altitudeToFovy(altitude);
  projectionParameters = getProjectionParameters({...});
}

super({
  // prev opts
  ...projectionParameters,
  focalDistance: altitude
});
```

Note that `orthographic` and `projectionMatrix` are already passed by `...opts`.

Comment (felixpalmer) [reply modules/core/src/viewports/web-mercator-viewport.js:81]: Fixed & removed `orthographic`

**Pull request merged**